Painless EqualsTest#testBranch(Not)?EqualsDefAndPrimitive fails on Windows

These two tests are currently failing on Windows only.

We have failures on 3.3 also.

```
=======================================
Elasticsearch Build Hamster says Hello!
=======================================
  Gradle Version        : 3.3
  OS Info               : Windows Server 2012 R2 6.3 (amd64)
  JDK Version           : Oracle Corporation 1.8.0_60 [Java HotSpot(TM) 64-Bit Server VM 25.60-b23]
  JAVA_HOME             : C:\Program Files\Java\jdk1.8.0_60
Incremental java compilation is an incubating feature.
.
.
.
[ant:junit4] JVM J3: stdout was not empty, see: C:\jenkins\workspace\ES Server 5.x on Windows 2012 R2\modules\lang-painless\build\testrun\test\temp\junit4-J3-20170409_202447_5686557531052253476795.sysout
Tests with failures:
  - org.elasticsearch.painless.EqualsTests.testBranchEqualsDefAndPrimitive
  - org.elasticsearch.painless.EqualsTests.testBranchNotEqualsDefAndPrimitive

Slow Tests Summary:
 12.69s | org.elasticsearch.painless.NeedsScoreTests
  7.50s | org.elasticsearch.painless.RemainderTests
  5.94s | org.elasticsearch.painless.ReservedWordTests
  4.99s | org.elasticsearch.painless.ScriptEngineTests
  4.94s | org.elasticsearch.painless.ScoreTests

[ant:junit4] JVM J0:     0.82 ..    15.46 =    14.63s
[ant:junit4] JVM J1:     0.82 ..    15.63 =    14.81s
[ant:junit4] JVM J2:     0.82 ..    15.89 =    15.06s
[ant:junit4] JVM J3:     0.82 ..    23.78 =    22.95s
[ant:junit4] Execution time total: 23 seconds
[ant:junit4] Tests summary: 46 suites, 749 tests, 2 failures, 1 ignored (1 assumption)
:modules:lang-painless:test FAILED
:modules:lang-painless:test (Thread[main,5,main]) completed. Took 23.858 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':modules:lang-painless:test'.
> There were test failures: 46 suites, 749 tests, 2 failures, 1 ignored (1 assumption) [seed: 4C7AA2AB6C0DF051]

* Try:
Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output.

BUILD FAILED

Total time: 49 mins 32.547 secs
Stopped 1 worker daemon(s).
```

--- a/modules/lang-painless/src/test/java/org/elasticsearch/painless/EqualsTests.java
+++ b/modules/lang-painless/src/test/java/org/elasticsearch/painless/EqualsTests.java
@@ -1,5 +1,3 @@
-package org.elasticsearch.painless;
-
 /*
  * Licensed to Elasticsearch under one or more contributor
  * license agreements. See the NOTICE file distributed with
@@ -19,7 +17,11 @@ package org.elasticsearch.painless;
  * under the License.
  */
 
-import org.apache.lucene.util.Constants;
+package org.elasticsearch.painless;
+
+import org.elasticsearch.test.ESTestCase;
+
+import static java.util.Collections.singletonMap;
 
 // TODO: Figure out a way to test autobox caching properly from methods such as Integer.valueOf(int);
 public class EqualsTests extends ScriptTestCase {
@@ -132,11 +134,13 @@ public class EqualsTests extends ScriptTestCase {
     }
     
     public void testBranchEqualsDefAndPrimitive() {
-        assumeFalse("test fails on Windows", Constants.WINDOWS);
-        assertEquals(true, exec("def x = 1000; int y = 1000; return x == y;"));
-        assertEquals(false, exec("def x = 1000; int y = 1000; return x === y;"));
-        assertEquals(true, exec("def x = 1000; int y = 1000; return y == x;"));
-        assertEquals(false, exec("def x = 1000; int y = 1000; return y === x;"));
+        /* This test needs an Integer that isn't cached by Integer.valueOf so we draw one randomly. We can't use any fixed integer because
+         * we can never be sure that the JVM hasn't configured itself to cache that Integer. It is sneaky like that. */
+        int uncachedAutoboxedInt = randomValueOtherThanMany(i -> Integer.valueOf(i) == Integer.valueOf(i), ESTestCase::randomInt);
+        assertEquals(true, exec("def x = params.i; int y = params.i; return x == y;", singletonMap("i", uncachedAutoboxedInt), true));
+        assertEquals(false, exec("def x = params.i; int y = params.i; return x === y;", singletonMap("i", uncachedAutoboxedInt), true));
+        assertEquals(true, exec("def x = params.i; int y = params.i; return y == x;", singletonMap("i", uncachedAutoboxedInt), true));
+        assertEquals(false, exec("def x = params.i; int y = params.i; return y === x;", singletonMap("i", uncachedAutoboxedInt), true));
     }
 
     public void testBranchNotEquals() {
@@ -150,11 +154,13 @@ public class EqualsTests extends ScriptTestCase {
     }
 
     public void testBranchNotEqualsDefAndPrimitive() {
-        assumeFalse("test fails on Windows", Constants.WINDOWS);
-        assertEquals(false, exec("def x = 1000; int y = 1000; return x != y;"));
-        assertEquals(true, exec("def x = 1000; int y = 1000; return x !== y;"));
-        assertEquals(false, exec("def x = 1000; int y = 1000; return y != x;"));
-        assertEquals(true, exec("def x = 1000; int y = 1000; return y !== x;"));
+        /* This test needs an Integer that isn't cached by Integer.valueOf so we draw one randomly. We can't use any fixed integer because
+         * we can never be sure that the JVM hasn't configured itself to cache that Integer. It is sneaky like that. */
+        int uncachedAutoboxedInt = randomValueOtherThanMany(i -> Integer.valueOf(i) == Integer.valueOf(i), ESTestCase::randomInt);
+        assertEquals(false, exec("def x = params.i; int y = params.i; return x != y;", singletonMap("i", uncachedAutoboxedInt), true));
+        assertEquals(true,  exec("def x = params.i; int y = params.i; return x !== y;", singletonMap("i", uncachedAutoboxedInt), true));
+        assertEquals(false, exec("def x = params.i; int y = params.i; return y != x;", singletonMap("i", uncachedAutoboxedInt), true));
+        assertEquals(true,  exec("def x = params.i; int y = params.i; return y !== x;", singletonMap("i", uncachedAutoboxedInt), true));
     }
 
     public void testRightHandNull() {

